oci/private/tarball.sh.tpl

```diff
@@ -1,37 +1,140 @@
+  #   "manifests": [
+  #     {
+  #       "mediaType": "application/vnd.oci.image.index.v1+json",
+  #       "size": 668,
+  #       "digest": "sha256:41981de3b7207f5260fd94fac77272218518d58a6335d843136d88d91341e3d9"
+  #     }
+  #   ]
+  # }
+  # Into:
+  # {
+  #   "schemaVersion": 2,
+  #   "manifests": [
+  #     {
+  #       "mediaType": "application/vnd.oci.image.index.v1+json",
+  #       "size": 668,
+  #       "digest": "sha256:41981de3b7207f5260fd94fac77272218518d58a6335d843136d88d91341e3d9",
+  #       "annotations": {
+  #         "org.opencontainers.image.ref.name": "repo-tag:1"
+  #       }
+  #     },
+  #     {
+  #       "mediaType": "application/vnd.oci.image.index.v1+json",
+  #       "size": 668,
+  #       "digest": "sha256:41981de3b7207f5260fd94fac77272218518d58a6335d843136d88d91341e3d9",
+  #       "annotations": {
+  #         "org.opencontainers.image.ref.name": "repo-tag:2"
+  #       }
+  #     }
+  #   ]
+  # }
+  repo_tags="${REPOTAGS[@]}" "${YQ}" -o json eval "(.manifests = ${MANIFEST_COPIES}) *d {\"manifests\": (env(repo_tags) | split \" \" | map {\"annotations\": {\"org.opencontainers.image.ref.name\": .}})}" "${INDEX_FILE}" > "${STAGING_DIR}/index.json"
+
+  tar -C "${STAGING_DIR}" -cf "${TARBALL_PATH}" index.json blobs oci-layout
+  exit 0
+fi
 
-MANIFEST_DIGEST=$(${YQ} eval '.manifests[0].digest | sub(":"; "/")' "${IMAGE_DIR}/index.json" | tr  -d '"')
 MANIFEST_BLOB_PATH="${IMAGE_DIR}/blobs/${MANIFEST_DIGEST}"
 
 CONFIG_DIGEST=$(${YQ} eval '.config.digest  | sub(":"; "/")' ${MANIFEST_BLOB_PATH})
 CONFIG_BLOB_PATH="${IMAGE_DIR}/blobs/${CONFIG_DIGEST}"
 
 LAYERS=$(${YQ} eval '.layers | map(.digest | sub(":"; "/"))' ${MANIFEST_BLOB_PATH})
 
-mkdir -p $(dirname "${BLOBS_DIR}/${CONFIG_DIGEST}")
-cp "${CONFIG_BLOB_PATH}" "${BLOBS_DIR}/${CONFIG_DIGEST}"
+cp_f_with_mkdir "${CONFIG_BLOB_PATH}" "${BLOBS_DIR}/${CONFIG_DIGEST}"
 
-for LAYER in $(${YQ} ".[]" <<< $LAYERS); do 
-    cp -f "${IMAGE_DIR}/blobs/${LAYER}" "${BLOBS_DIR}/${LAYER}.tar.gz"
+for LAYER in $(${YQ} ".[]" <<< $LAYERS); do
+  cp_f_with_mkdir "${IMAGE_DIR}/blobs/${LAYER}" "${BLOBS_DIR}/${LAYER}.tar.gz"
 done
 
-# Replace newlines (unix or windows line endings) with % character.
-# We can't pass newlines to yq due to https://github.com/mikefarah/yq/issues/1430 and
-# we can't update YQ at the moment because structure_test depends on a specific version:
-# see https://github.com/bazel-contrib/rules_oci/issues/212
-repo_tags="$(tr -d '\r' < "${TAGS_FILE}" | tr '\n' '%')" \
+repo_tags="${REPOTAGS[@]}" \
 config="blobs/${CONFIG_DIGEST}" \
 layers="${LAYERS}" \
 "${YQ}" eval \
-        --null-input '.[0] = {"Config": env(config), "RepoTags": "${repo_tags}" | envsubst | split("%") | map(select(. != "")) , "Layers": env(layers) | map( "blobs/" + . + ".tar.gz") }' \
+        --null-input '.[0] = {"Config": env(config), "RepoTags": "${repo_tags}" | envsubst | split(" ") | map(select(. != "")) , "Layers": env(layers) | map( "blobs/" + . + ".tar.gz") }' \
```

Comment:
this change seems unrelated. can we take that back?

Reply:
I've done this, but I don't think it's how we should leave this long-term; I see two issues:
1. Now the two codepaths handle repotags differently, which may become a source of drift/bugs in the future.
2. I think the "just have bash split things into an array and use the array" approach seems significantly simpler than "read the file, replace newlines with percents, have a 4-line comment explaining why we're doing so, then split again on perecents".

Happy to merge this with the two ways, and maybe follow up unifying them in the future, if that'd make sense? Or just reverting the "Revert repo_tags change" commit in this PR and accepting needting to update other PRs for the conflict? Whatever works best for y'all.

Reply:
Prefactor PR: https://github.com/bazel-contrib/rules_oci/pull/401

Reply:
Merged and rebased - PTAL! Thanks!